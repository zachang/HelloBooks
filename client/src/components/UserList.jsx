import React from 'react';
import PropTypes from 'react-proptypes';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Pagination } from 'react-materialize';
import { getUserAction } from '../actions/userAction';
import UserRow from './user/UserRow.jsx';

/**
 * UserList class declaration
 * @class UserList
 * @extends {React.Component}
 */
export class UserList extends React.Component {
  /**
   * class constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      errors: null,
      pageCount: null,
      limit: 15
    };
  }

  /**
   * @method componentWillMount
   * @return {void} void
   */
  componentWillMount() {
    this.props.getUserAction(this.state.limit, 0);
  }

  /**
   * @method componentWillReceiveProps
   * @param {object} nextProps - nextProps
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.userState.success === false) {
      this.setState({ errors: nextProps.userState.errors });
    }

    if (this.state.pageCount !== nextProps.userState.pageCount) {
      this.setState({ pageCount: nextProps.userState.pageCount });
    }
  }

  /**
   * Renders UserList component
   * @return {XML} JSX
   */
  render() {
    return (
      <div className='row'>
        <div className='section'>
          <h4 style={{ marginTop: '7%' }}>Borrowed Books</h4>
        </div>
        <div className='divider' style={{ width: '100%', marginTop: '-2.5%', marginBottom: '3.5%' }}></div>

        <div className='row'>
          <table className='bordered highlight'>
            <thead className='black white-text'>
              <tr>
                <th>Name</th>
                <th>Username</th>
                <th>Phone</th>
                <th>email</th>
                <th>borrow History</th>
              </tr>
            </thead>

            <tbody>
              { this.props.userState.users.map((user, i) =>
                <UserRow
                  key={i}
                  user={user}
                />
              )}
            </tbody>
          </table>
        </div>

        <div className='row'>
          {
            ((this.state.pageCount) ?
              <Pagination
                items={this.state.pageCount}
                onSelect={(page) => {
                  const offset = (page - 1) * this.state.limit;
                  this.props.getUserAction(this.state.limit, offset);
                }} /> : '')
          }
        </div>
      </div>
    );
  }
}

UserList.propTypes = {
  userState: PropTypes.object.isRequired,
  tokenState: PropTypes.object.isRequired,
  getUserAction: PropTypes.func.isRequired,
};
const mapStateToProps = state => ({
  userState: state.userReducer,
  tokenState: state.tokenReducer
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ getUserAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(UserList);