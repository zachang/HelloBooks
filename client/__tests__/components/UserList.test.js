import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow, mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { UserList } from '../../src/components/UserList.jsx';


describe('<UserList/>', () => {
  // create a spy function for getUserAction
  const getUserAction = sinon.spy();
  // spy on componentWillReceiveProps of Borrowed Page
  sinon.spy(UserList.prototype, 'componentWillReceiveProps');
  // spy on componentWillMount of Borrowed Page
  sinon.spy(UserList.prototype, 'componentWillMount');

  const errors = null;
  const props = {
    userState: {
      success: null,
      errors: null,
      fails: null,
      users: ['mmmmmmm'],
      user: [],
      pageCount: null,
    },
    getUserAction,
    errors,
    pageCount: null,
    limit: 15
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<UserList {...props} />);
    shallowComponent.setState({
      pageCount: 1
    });
  });

  it('renders <UserList /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it(`should check if the 'UserRow' is defined`, () => {
    expect(shallowComponent.find('UserRow').length).toBe(1);
  });

  it('calls componentWillReceiveProps when paginating', () => {
    props.pageCount = 1;
    props.userState.pageCount = 2;
    shallowComponent.setState({
      pageCount: props.userState.pageCount
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(UserList.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when error in user displays', () => {
    props.userState.success = false;
    props.userState.errors = 'failed';
    shallowComponent.setState({
      errors: props.userState.errors
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(UserList.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it(`should check if the 'Pagination' is defined`, () => {
    expect(shallowComponent.find('Pagination').length).toBe(1);
    shallowComponent.find('Pagination').simulate('select');
  });

  it(`should check if the 'Pagination' is defined`, () => {
    shallowComponent.instance().componentWillMount();
    expect(UserList.prototype.componentWillMount.called).toEqual(true);
  });
});
