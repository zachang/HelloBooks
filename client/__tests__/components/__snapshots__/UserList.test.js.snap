// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList/> renders <UserList /> component 1`] = `
<div
  className="row"
>
  <div
    className="section"
  >
    <h4
      style={
        Object {
          "marginTop": "7%",
        }
      }
    >
      Borrowed Books
    </h4>
  </div>
  <div
    className="divider"
    style={
      Object {
        "marginBottom": "3.5%",
        "marginTop": "-2.5%",
        "width": "100%",
      }
    }
  />
  <div
    className="row"
  >
    <table
      className="bordered highlight"
    >
      <thead
        className="black white-text"
      >
        <tr>
          <th>
            Name
          </th>
          <th>
            Username
          </th>
          <th>
            Phone
          </th>
          <th>
            email
          </th>
        </tr>
      </thead>
      <tbody>
        <UserRow
          key="0"
          user="mmmmmmm"
        />
      </tbody>
    </table>
  </div>
  <div
    className="row"
  >
    <Pagination
      activePage={1}
      items={1}
      onSelect={[Function]}
    />
  </div>
</div>
`;
